Remotion/Web/Development.WebTesting/Utilities/ActionsWithWaitSupport.cs

```diff
@@ -1,88 +0,0 @@
-﻿// This file is part of the re-motion Core Framework (www.re-motion.org)
-// Copyright (c) rubicon IT GmbH, www.rubicon.eu
-// 
-// The re-motion Core Framework is free software; you can redistribute it 
-// and/or modify it under the terms of the GNU Lesser General Public License 
-// as published by the Free Software Foundation; either version 2.1 of the 
-// License, or (at your option) any later version.
-// 
-// re-motion is distributed in the hope that it will be useful, 
-// but WITHOUT ANY WARRANTY; without even the implied warranty of 
-// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
-// GNU Lesser General Public License for more details.
-// 
-// You should have received a copy of the GNU Lesser General Public License
-// along with re-motion; if not, see http://www.gnu.org/licenses.
-// 
-using System;
-using JetBrains.Annotations;
-using OpenQA.Selenium;
-using OpenQA.Selenium.Interactions;
-using OpenQA.Selenium.Support.UI;
-using Remotion.Utilities;
-
-namespace Remotion.Web.Development.WebTesting.Utilities
-{
-  /// <summary>
-  /// Enhanced version of Selenium's <see cref="Actions"/> class. It enables us to add WaitFor predicates after actions.
-  /// </summary>
-  public class ActionsWithWaitSupport : Actions
```

Comment:
Why was this type removed? I do not see any usages in re-motion but it is problably introduced to help something downstream. Unless keeping this type would be harmful (i.e. using it with Selenium 4.5 would break things) we should not delete it at this time. 

Note; if we know that it's unused, we can make a task for marking it as obsolete in the next major release.

Reply:
The problem here is that parts of the `protected` `Actions` APIs are no longer available. In this case, the `AddAction` can no longer be accessed. We need to decide what to do with the class. The delete was just to remember that there is something left to do :D

Reply:
Okay, I've checked in the project sources and found usages of ActionsWithWaitSupport -> the assumtion that everything is already done is not true, invalidating the estimate. => should we add a separate task to solve this? The impleemntation will still be in the main feature branch, but we may want to separate the estimate/planning for it.

Reply:
Note: it may be (please verify) that there's only one of dependency on this replicated several times. If there's a good alternative to solving the dependency, we don't need to implement this in re-motion the same way as it was done before.

Reply:
`Actions` uses an `ActionBuilder` to append all the actions and then build and execute them. This builder is private and there is no way to access it. I don't quite understand why there is no extensibility on this type provided by Selenium.

My first approach would be to generate a delegate that can access the private field. Then we only need to change the `AddAction` to `GetBuilder().AddAction`. The delegate can be created once as part of static initialization and get pretty close to a normal field access (21t vs 19t). Here is an example code of that in action using a `StringBuilder` instead of `ActionBuilder` as I did not want to add the Selenium reference to my test project:

```c#
public class MyActions : Actions
{
  private static Func<Actions, StringBuilder> s_getter;

  static MyActions()
  {
    var dynamicMethod = new DynamicMethod(string.Empty, typeof(StringBuilder), new []{typeof(Actions)}, true);
    var ilGenerator = dynamicMethod.GetILGenerator();
    ilGenerator.Emit(OpCodes.Ldarg_0);
    ilGenerator.Emit(OpCodes.Ldfld, typeof(Actions).GetField("_stringBuilder", BindingFlags.NonPublic | BindingFlags.Instance));
    ilGenerator.Emit(OpCodes.Ret);

    s_getter = (Func<Actions, StringBuilder>)dynamicMethod.CreateDelegate(typeof(Func<Actions, StringBuilder>));
  }
  
  public void SayGoodbye()
  {
    s_getter(this).AppendLine("GOODBYE");
  }
}
```

`SayGoodbye` would be our `WaitFor` in this example. The biggest benefit of this approach is that we don't break any existing code. If we introduce the delegate on a helper class or introduce a base class we can help consumers that run into similar problems.

Reply:
We don't need to do IL-Code here (and should use Expression instead, if we really want to), simply caching the field is more than enough in the webtest code.

Add an obsolete attribute (and breaking change doc that it is now obsolete) to this type and a follow-up ticket to remove it in re-motion v4.0. Please verify, but if I'm correct, there's other (also simple) ways to solve this ad-hoc on a project level and it removes our repsonsibility for working around Selenium design desicions.